Segregate celo transaction logic

piersy opened the pull request:
This PR attempts to minimise intermingling of celo transaction processing with the logic within the op geth codebase. It does this by calling out to celo specific functions that are held in new files and thereby lessens the mixing/interleaving of celo and op-geth logic.

Going forward having a less intertwined diff should reduce merge conflicts and thereby speed up pulling in upstream changes as well as making it less likely that bugs arise out of the merging process.

It's useful to view the diff created by the initial addition of celo transaction support and the diff with these changes incorporated to see how they differ in modifications to op-geth owned files.

It's useful to view our total diff on the types package before and after these changes to get an idea of how much this code disentangles us from the types package.

The following two commands let you see our modifications to the existing source of the types package, without the clutter of added files.

```
d --diff-filter=M 33cb3315f498f219f95c9f3e42aa1f3742fd9d18 celo4 -- ./core/types
d --diff-filter=M 33cb3315f498f219f95c9f3e42aa1f3742fd9d18 piersy/segregate-celo-transactions -- ./core/types
```

Note that celo receipt logic has not been disentangled as part of this PR and it is noticably present in the types package modifications, I think it could make sense to include segregating the receipt logic in this PR, but it wasn't my initial area of focus.

~~Also note that as a final step I moved all the celo modifications into one file `celo_extensions.go` to see how that looked. It seems like possibly an easier way to keep track of our changes than having them scattered through many randomly named new files. Obviously that will only work well up to a certain amount of changes and maybe we already have enough changes here in the types package to justify separate celo files but I thought I'd introduce the idea.~~ Decided to follow @mcortesi's [suggestion](https://github.com/celo-org/op-geth/pull/115#discussion_r1578409952)

Closes - https://github.com/celo-org/celo-blockchain-planning/issues/230

piersy committed:
Move tx marshalling to own file as much as possible

piersy committed:
Remove feeCurrency notion from op code

piersy committed:
Segregate celo transaction types

piersy committed:
Seggregate celoDecodeTyped

piersy committed:
Segregate celo tx singing code

piersy committed:
Move CompareWithRates into legacypool package

There was no need for it to be in the types package

piersy committed:
Consolidate celo changes in one file

mcortesi commented on core/txpool/legacypool/list.go:
```diff
@@ -503,7 +503,7 @@ func (h *priceHeap) Less(i, j int) bool {
 
 func (h *priceHeap) cmp(a, b *types.Transaction) int {
-	return types.CompareWithRates(a, b, h.ratesAndFees)
+	return compareWithRates(a, b, h.ratesAndFees)
```
i like this since it's not introdcing a package dependency, respects public/private methods, etc..

piersy replied:
There didn't seem to be any reason for CompareWithRates to be in the types package.

mcortesi commented on core/types/deposit_tx.go:
```diff
@@ -101,5 +101,3 @@ func (tx *DepositTx) encode(b *bytes.Buffer) error {
 func (tx *DepositTx) decode(input []byte) error {
 	return rlp.DecodeBytes(input, tx)
 }
-
```
this works, moving everything to a new place. The alternative which is also not bad, is as a rule of thumb all modification go to the end of the file. That simplifies merging/diffs

piersy replied:
Adding to the end of the file still results in conflicts if something else is added to the end of the file, so I think moving to a separate file is the best option.

Notice also in the example below that if you resolved it manually you'd probably end up with `MakeSignerOp` at the end of the file, so now the celo change would not be at the end, and therefore be at increased risk of merge conflicts. 

I tried adding a function to the end of the same file from 2 branches and this is the result when merging.

```
diff --cc core/types/transaction_signing.go
index 56cc40f37,1d9045f71..000000000
--- a/core/types/transaction_signing.go
+++ b/core/types/transaction_signing.go
@@@ -601,16 -601,25 +601,42 @@@ func deriveChainId(v *big.Int) *big.In
    v = new(big.Int).Sub(v, big.NewInt(35))
    return v.Div(v, big.NewInt(2))
  }
++<<<<<<< HEAD
 +
 +// deriveChainId derives the chain id from the given v parameter
 +func deriveChainIdCelo(v *big.Int) *big.Int {
 +  if v.BitLen() <= 64 {
 +      v := v.Uint64()
 +      if v == 27 || v == 28 {
 +          return new(big.Int)
 +      }
 +      return new(big.Int).SetUint64((v - 35) / 2)
 +  }
 +  v = new(big.Int).Sub(v, big.NewInt(35))
 +  return v.Div(v, big.NewInt(2))
 +}
++||||||| b40e751e8
++=======
+
+ // MakeSigner returns a Signer based on the given chain config and block number.
+ func MakeSignerOp(config *params.ChainConfig, blockNumber *big.Int, blockTime uint64) Signer {
+   var signer Signer
+   switch {
+   case config.IsCancun(blockNumber, blockTime) && !config.IsOptimism():
+       signer = NewCancunSigner(config.ChainID)
+   case config.IsCel2(blockTime):
+       signer = NewCel2Signer(config.ChainID)
+   case config.IsLondon(blockNumber):
+       signer = NewLondonSigner(config.ChainID)
+   case config.IsBerlin(blockNumber):
+       signer = NewEIP2930Signer(config.ChainID)
+   case config.IsEIP155(blockNumber):
+       signer = NewEIP155Signer(config.ChainID)
+   case config.IsHomestead(blockNumber):
+       signer = HomesteadSigner{}
+   default:
+       signer = FrontierSigner{}
+   }
+   return signer
+ }
++>>>>>>> piersy/merge-test3
```

mcortesi commented on core/types/celo_extensions.go:
```diff
@@ -0,0 +1,319 @@
+)
+
+// Returns the fee currency of the transaction if there is one.
+func (tx *Transaction) FeeCurrency() *common.Address {
```
i think having a single celo_extensions.go file could become quite messy.

You have a collection of methods, sometimes, extending an existing interface, sometimes adding some filed, etc..

Maybe a good way to do this is to follow a rule like:

* If you need to modify some op-geth file xxx.go, then extension methods go to celo_xxx.go file

In our case, things that we added to transaction.go, are added to celo_transaction.go

Then we have other changes that are NEW files... like new celo envelop txs than can have its own file

piersy replied:
Sounds good, I will revert the change to put everything in one file and then re-organise as you suggest. Having everything in one file also increases chances of merge conflicts for our own updates.

mcortesi commented on core/types/transaction.go:
```diff
@@ -205,14 +200,17 @@ func (tx *Transaction) decodeTyped(b []byte) (TxData, error) {
 		return nil, errShortTypedTx
 	}
+
+	if inner, isCelo, err := celoDecodeTyped(b); isCelo {
```
i like the strategy:

Add an `if` at the begginign of the method, that deals with "everything celo". Works better when that's not a single thing, but many. Like many different tx that needs to be handled

piersy replied:
I'd say the overhead is better amortised when it's many things. But even with one thing not having modifications in the middle of the switch statement will likely reduce the chance of conflicts because all new transaction additions will modify the switch statement.

mcortesi commented on core/types/transaction_signing.go:
```diff
@@ -38,12 +38,15 @@ type sigCache struct {
 // MakeSigner returns a Signer based on the given chain config and block number.
 func MakeSigner(config *params.ChainConfig, blockNumber *big.Int, blockTime uint64) Signer {
+
+	if signer, isCelo := makeCeloSigner(config, blockNumber, blockTime); isCelo {
```
IDEA: what if we make this the strategy for extending OP-stack, and renane methods to be more general like `makeSignerHook(...)`

And so we would be defining hook points that can be used with any op stack, and have a default impl of `nil`... if we do this nicely we ca nprobably get this merged upstream

piersy replied:
Sounds good to me. So you'd put the default impl in a separate file named `transaction_signing_hooks.go` ?

But maybe this is a strategy for extending ethereum rather than just for the op stack, so we could try to get it in up up stream?

piersy replied:
Although maybe Hook is not the best postfix, because I see a lot of xxxHook things already in the codbase, it would be nice to use something distinct to aid finding all the hookpoints.

mcortesi replied:
yeah, don't mind the name... more on something that doesn't say celo, because it's the extension point to any other op chain.

mcortesi reviewed (commented):
I like the idea, left some comments

piersy committed:
Fix linter error for unnecessary leading newlines

piersy committed:
Revert "Consolidate celo changes in one file"

This reverts commit a37e58f286988b325b3e0d12aaea206dd6081e84.

piersy committed:
Rename celo_extensions.go to celo_transaction.go

piersy committed:
Switch to generic naming for custom functions

This lays some groundwork for this approach to be used by projects other
than celo.

piersy committed:
Add comments for extension functions

piersy committed:
Fix signer creation logic

We wern't asserting that the config was "Optimism" which means in our
case celo, which allowed us to return celo signers for non celo
situations.

piersy committed:
Handle errors in transfer tests

piersy committed:
Fix transfer tests

The tests were not being started with a valid CeL2 config.

palango commented on params/config.go:
```diff
@@ -678,7 +678,7 @@ func (c *ChainConfig) IsInterop(time uint64) bool {
 
 func (c *ChainConfig) IsCel2(time uint64) bool {
-	return isTimestampForked(c.Cel2Time, time)
+	return c.IsOptimism() && isTimestampForked(c.Cel2Time, time)
```
In which case was this wrong?

When we added the hardfork flag we thought of it as a feature flag for celo specific behaviour. So from my perspective it only enables new transaction types (and more), but is not inherently bound to optimism (that might change though)

piersy replied:
My understanding was that the transition to cel2 would mean that we are also optimism, because at that point we would be using the op-stack and we would be setting the optimism config.

How would we make the transition to be cel2 with the optimism codebase but not be optimism?

piersy committed:
Update assertion to match maxFeePerGas and baseFee

The values changed with the introduction of the Optimism config

piersy committed:
Fix test by setting cel2 config

piersy commented:
Closing this since most of this has now made its way into op geth in other PRs

* #165
* #159 
* #150
